### ISSUE TITLE
Kubelet does not start after node reboot

### ISSUE THREAD

**Issue by avalluri** [clearlinux/distribution#1004]
On Clear Linux 3020, the kubelet service fails to start once after a node reboot. It fails to connect to docker though docker is running.

```
$ journalctl -au kubelet

Jul 12 09:30:11 pmem-csi-clear-govm-worker1 kubelet[391]: I0712 09:30:11.647366     391 kubelet.go:306] Watching apiserver
Jul 12 09:30:11 pmem-csi-clear-govm-worker1 kubelet[391]: E0712 09:30:11.647604     391 file.go:98] Unable to read config path "/etc/kubernetes/manifests": path does not exist, ignoring
Jul 12 09:30:11 pmem-csi-clear-govm-worker1 kubelet[391]: E0712 09:30:11.647672     391 file_linux.go:61] Unable to read config path "/etc/kubernetes/manifests": path does not exist, ignoring
Jul 12 09:30:11 pmem-csi-clear-govm-worker1 kubelet[391]: I0712 09:30:11.651254     391 client.go:75] Connecting to docker on unix:///var/run/docker.sock
Jul 12 09:30:11 pmem-csi-clear-govm-worker1 kubelet[391]: I0712 09:30:11.651303     391 client.go:104] Start docker client with request timeout=2m0s
Jul 12 09:30:11 pmem-csi-clear-govm-worker1 kubelet[391]: F0712 09:30:11.651868     391 server.go:273] failed to run Kubelet: failed to create kubelet: failed to get docker version: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
Jul 12 09:30:11 pmem-csi-clear-govm-worker1 systemd[1]: kubelet.service: Main process exited, code=exited, status=255/EXCEPTION
Jul 12 09:30:11 pmem-csi-clear-govm-worker1 systemd[1]: kubelet.service: Failed with result 'exit-code'.
Jul 12 09:30:21 pmem-csi-clear-govm-worker1 systemd[1]: kubelet.service: Service RestartSec=10s expired, scheduling restart.
Jul 12 09:30:21 pmem-csi-clear-govm-worker1 systemd[1]: kubelet.service: Scheduled restart job, restart counter is at 6.
Jul 12 09:30:21 pmem-csi-clear-govm-worker1 systemd[1]: Stopped kubelet: The Kubernetes Node Agent.
Jul 12 09:30:21 pmem-csi-clear-govm-worker1 systemd[1]: kubelet.service: Start request repeated too quickly.
Jul 12 09:30:21 pmem-csi-clear-govm-worker1 systemd[1]: kubelet.service: Failed with result 'exit-code'.
```

**Comment by pohly** [comment]
See also issue #891 - another issue when rebooting.

**Comment by pohly** [comment]
The same happens when using CRI-O. I've tested on Clear Linux  30320.
```

Jul 12 15:19:47 pmem-csi-clear-govm-lvm-crio-worker1 kubelet[358]: E0712 15:19:47.177950     358 remote_runtime.go:81] Version from runtime service failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial unix /var/run/crio/crio.sock: connect: no such file or directory"
Jul 12 15:19:47 pmem-csi-clear-govm-lvm-crio-worker1 kubelet[358]: E0712 15:19:47.178094     358 kuberuntime_manager.go:191] Get runtime version failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial unix /var/run/crio/crio.sock: connect: no such file or directory"
Jul 12 15:19:47 pmem-csi-clear-govm-lvm-crio-worker1 kubelet[358]: F0712 15:19:47.178142     358 server.go:273] failed to run Kubelet: failed to create kubelet: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial unix /var/run/crio/crio.sock: connect: no such file or directory"
Jul 12 15:19:47 pmem-csi-clear-govm-lvm-crio-worker1 systemd[1]: kubelet.service: Main process exited, code=exited, status=255/EXCEPTION
Jul 12 15:19:47 pmem-csi-clear-govm-lvm-crio-worker1 systemd[1]: kubelet.service: Failed with result 'exit-code'.
Jul 12 15:19:57 pmem-csi-clear-govm-lvm-crio-worker1 systemd[1]: kubelet.service: Service RestartSec=10s expired, scheduling restart.
Jul 12 15:19:57 pmem-csi-clear-govm-lvm-crio-worker1 systemd[1]: kubelet.service: Scheduled restart job, restart counter is at 6.
Jul 12 15:19:57 pmem-csi-clear-govm-lvm-crio-worker1 systemd[1]: Stopped kubelet: The Kubernetes Node Agent.
```

While I was investigating and rebooted once more, it got worse: ssh apparently also didn't come up again. At least I only get "connection refused" for port 22 although the machine itself responds to pings.

**Comment by pohly** [comment]
ssh did come up eventually - just slower than expected.

**Comment by hnanni** [comment]
I'll need help to reproduce this.
I'm using cri-o in

```
root@kube2~ # swupd info
Installed version: 30340
```

The master observes the reboot and then the machine coming back:
```
root@master~ # kubectl get nodes -o wide
NAME     STATUS   ROLES    AGE   VERSION      EXTERNAL-IP   OS-IMAGE         KERNEL-VERSION   CONTAINER-RUNTIME
kube1    Ready    <none>   56m   v1.15.0      <none>        Clear Linux OS   5.1.17-363.kvm   cri-o://1.14.4
kube2    Ready    <none>   55m   v1.15.0      <none>        Clear Linux OS   5.1.17-363.kvm   cri-o://1.14.4
kube3    Ready    <none>   54m   v1.15.0      <none>        Clear Linux OS   5.1.17-363.kvm   cri-o://1.14.4
master   Ready    master   57m   v1.15.0      <none>        Clear Linux OS   5.1.17-363.kvm   cri-o://1.14.4
root@master~ # kubectl get nodes -o wide
NAME     STATUS     ROLES    AGE   VERSION      EXTERNAL-IP   OS-IMAGE         KERNEL-VERSION   CONTAINER-RUNTIME
kube1    Ready      <none>   58m   v1.15.0   <none>        Clear Linux OS   5.1.17-363.kvm   cri-o://1.14.4
kube2    NotReady   <none>   57m   v1.15.0   <none>        Clear Linux OS   5.1.17-363.kvm   cri-o://1.14.4
kube3    Ready      <none>   56m   v1.15.0   <none>        Clear Linux OS   5.1.17-363.kvm   cri-o://1.14.4
master   Ready      master   59m   v1.15.0   <none>        Clear Linux OS   5.1.17-363.kvm   cri-o://1.14.4
root@master~ # kubectl get nodes -o wide
NAME     STATUS   ROLES    AGE   VERSION      EXTERNAL-IP   OS-IMAGE         KERNEL-VERSION   CONTAINER-RUNTIME
kube1    Ready    <none>   59m   v1.15.0      <none>        Clear Linux OS   5.1.17-363.kvm   cri-o://1.14.4
kube2    Ready    <none>   58m   v1.15.0      <none>        Clear Linux OS   5.1.17-363.kvm   cri-o://1.14.4
kube3    Ready    <none>   57m   v1.15.0      <none>        Clear Linux OS   5.1.17-363.kvm   cri-o://1.14.4
master   Ready    master   59m   v1.15.0      <none>        Clear Linux OS   5.1.17-363.kvm   cri-o://1.14.4

```

This is the kubelet.service after a reboot:

```
root@kube2~ # systemctl status kubelet
● kubelet.service - kubelet: The Kubernetes Node Agent
   Loaded: loaded (/usr/lib/systemd/system/kubelet.service; enabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/kubelet.service.d
           └─proxy.conf
   Active: active (running) since Fri 2019-07-12 14:15:43 CDT; 1min 46s ago
     Docs: http://kubernetes.io/docs/
 Main PID: 309 (kubelet)
    Tasks: 0 (limit: 4720)
   Memory: 23.4M
   CGroup: /system.slice/kubelet.service
           ‣ 309 /usr/bin/kubelet --runtime-cgroups=/sys/fs/cgroup/systemd/system.slice --kubelet-cgroups=/sys/fs/cgroup/systemd/system.slice --bootstrap-kubeconfig=/etc/kubernetes/bootstrap>

Jul 12 14:16:04 kube2 kubelet[309]: I0712 14:16:04.138613     309 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "canal-token-5c8k8" (UniqueName: "kub>
Jul 12 14:16:04 kube2 kubelet[309]: I0712 14:16:04.138704     309 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "local-bin" (UniqueName: "kubernetes.>
Jul 12 14:16:04 kube2 kubelet[309]: I0712 14:16:04.138822     309 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "lib-modules" (UniqueName: "kubernete>
Jul 12 14:16:04 kube2 kubelet[309]: I0712 14:16:04.138960     309 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "flannel-cfg" (UniqueName: "kubernete>
Jul 12 14:16:04 kube2 kubelet[309]: I0712 14:16:04.139101     309 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "kata-artifacts" (UniqueName: "kubern>
Jul 12 14:16:04 kube2 kubelet[309]: I0712 14:16:04.139201     309 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "containerd-conf" (UniqueName: "kuber>
Jul 12 14:16:04 kube2 kubelet[309]: I0712 14:16:04.139454     309 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "var-run-calico" (UniqueName: "kubern>
Jul 12 14:16:04 kube2 kubelet[309]: I0712 14:16:04.139545     309 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "cni-bin-dir" (UniqueName: "kubernete>
Jul 12 14:16:04 kube2 kubelet[309]: I0712 14:16:04.139613     309 reconciler.go:150] Reconciler: start to sync state
Jul 12 14:16:04 kube2 kubelet[309]: I0712 14:16:04.403063     309 kubelet_node_status.go:72] Attempting to register node kube2
```

**Comment by hnanni** [comment]
for the ssh issue, it might be the case this is relevant:
https://clearlinux.org/news-blogs/my-clear-linux-system-just-stopped-responding-ssh

**Comment by pohly** [comment]
Humberto Nanni <notifications@github.com> writes:
> I'll need help to reproduce this.

It seems to be a race condition and might not show up on any of your
systems. A test would be to install Kubernetes and then reboot 100
times.

Let's turn this around: how does the system ensure that kubelet is
started after the container runtime system that it depends on?

Should that be covered for the default case (Kubernetes + CRI-O), or is
that something that needs to be configured by the admin by extending the
systemd unit for kubelet?

The best solution would be to use systemd socket activation for CRI-O
and Docker, then it would work automatically. This is not currently
configured, is it?

**Comment by amygilliam1** [comment]
@hnanni , any update on this issue?

**Comment by hnanni** [comment]
No @amygilliam1 , not yet. I have not implemented stress tests that may explore the posibility of a reace condition, neither have worked on tweaking the initialization of kubelet to force it waiting for the container runtime. Please letme know if you need I raise the priority of those tasks

**Comment by pohly** [comment]
I've not seen this problem occur with recent Clear Linux (30900), but that doesn't mean that it is gone (it's a timing problem after all).

This unit dependency is something that can be configured by whatever script sets up Kubernetes on Clear Linux. For the test cluster in PMEM-CSI, I created a PR for it: https://github.com/intel/pmem-csi/pull/395

@jascott1, @CraigSterrett, @ganeshmaharaj: perhaps https://github.com/clearlinux/cloud-native-setup should do the same?

**Comment by hnanni** [comment]
has this been address by the fix in https://github.com/clearlinux/cloud-native-setup/issues/127 ?

**Comment by pohly** [comment]
That issue looks like it's about something else.

Personally I don't mind closing this issue here with the explanation that configuring the systemd unit dependency is something that has to be done by the Kubernetes setup script, but for that we need to hear from @jascott1, @CraigSterrett and/or @ganeshmaharaj.

**Comment by avalluri** [comment]
I am not sure what fixed the issue, but I couldn't reproduce on latest(31090) image. I am ok to close this.

**Comment by pohly** [comment]
Amarnath Valluri <notifications@github.com> writes:
> I am not sure what fixed the issue, but I couldn't reproduce on
> latest(31090) image. I am ok to close this.

Did you test with PMEM-CSI? I fixed it there with
https://github.com/intel/pmem-csi/pull/395.

**Comment by avalluri** [comment]
> Amarnath Valluri <notifications@github.com> writes:
> I am not sure what fixed the issue, but I couldn't reproduce on latest(31090) image. I am ok to close this.
> Did you test with PMEM-CSI? I fixed it there with [intel/pmem-csi#395](https://github.com/intel/pmem-csi/pull/395).

Yes, I am aware of your change. I tested by reverting that change.
